File: Ahorn/entities/KrackoBoss.jl
module KrackoBoss

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/KrackoBoss" KrackoBoss(x::Integer, y::Integer, health::Integer=900, maxHealth::Integer=900)

const placements = Ahorn.PlacementDict(
    "kracko_boss" => Ahorn.EntityPlacement(KrackoBoss)
)

function Ahorn.selection(entity::KrackoBoss)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::KrackoBoss, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: Loenn/entities/kracko_boss.lua
local drawableSprite = require("structs.drawable_sprite")
local utils = require("utils")

local krackoBoss = {}

krackoBoss.name = "MaggyHelper/KrackoBoss"
krackoBoss.depth = -8500
krackoBoss.justification = {0.5, 0.5}

krackoBoss.placements = {
    {
        name = "kracko_boss",
        data = {
            health = 900,
            maxHealth = 900,
    }
    }
}

krackoBoss.fieldInformation = {
    health = {
        fieldType = "integer",
        minimumValue = 1
    },
    maxHealth = {
        fieldType = "integer",
        minimumValue = 1
    }
}

krackoBoss.fieldOrder = {
    "x", "y",
    "health",
    "maxHealth"
}

function krackoBoss.sprite(room, entity)
    local textures = {
        "bosses/kirby/kracko/idle00",
        "characters/Kglobal::Player/sitDown00"
    }

    for _, texture in ipairs(textures) do
        local ok, sprite = pcall(drawableSprite.fromTexture, texture, entity)
        if ok and sprite then
            sprite:setJustification(0.5, 0.5)
            return {sprite}
        end
    end

    return {}
end

function krackoBoss.selection(room, entity)
    return utils.rectangle(entity.x - 24, entity.y - 24, 48, 48)
end

return krackoBoss
